### DISCUSSION TITLE
Understanding how each input contributes to the output (e.g. using SHAP)

### DISCUSSION THREAD

**Discussion by LennardLux** [neuralhydrology/neuralhydrology#148] [Q&A]
Hi, 
I have trained some models now and achieved some acceptable results. Now I want to understand how much and in which way each input contributes to the output variable. I stumbled upon a publication by [Wunsch et. al (2022)](https://www.nature.com/articles/s41467-022-28770-2/figures/8) in which [SHapley Additive exPlanations (SHAP)](https://shap.readthedocs.io/en/latest/) have been used to visualize this. I find that the figure SHAP produces is really intuitive and easy to read. The authors used Tensorflow instead of Pytorch. Has anyone tried using SHAP using the cudaLSTM-Model and could share some experiences? I am really new to XAI-Approaches and don't have a good overview about what's out there and what's compatible with what. I would greatly appreciate any recommendations or hints!

![Screenshot 2023-10-12 122039](https://github.com/neuralhydrology/neuralhydrology/assets/107072862/62ccd190-3ffb-467d-b92f-27e4458b370a)
Wunsch, A., Liesch, T. & Broda, S. Deep learning shows declining groundwater levels in Germany until 2100 due to climate change. Nat Commun 13, 1221 (2022). https://doi.org/10.1038/s41467-022-28770-2

**Comment by gauchm** [comment]
Hi @LennardLux,
I don't have any experience in working with SHAP myself, so I don't have much to share. But from [this example](https://shap.readthedocs.io/en/latest/example_notebooks/image_examples/image_classification/PyTorch%20Deep%20Explainer%20MNIST%20example.html) on their website it looks like they support analyzing PyTorch models in the same way as Wunsch et al. did in their [published code](https://github.com/AndreasWunsch/Long-Term-GWL-Simulations/blob/main/Code/04_check_plausibility_SHAP.py#L146). 

If you are more concerned about how you would load a trained model and get the training/validation data to start your analyses, maybe [this tutorial](https://neuralhydrology.readthedocs.io/en/latest/tutorials/inspect-lstm.html) in our docs helps.

**Reply by LennardLux** [reply]
Hi @gauchm, thanks for you response. I did some digging and stumbled upon [#213](https://github.com/shap/shap/issues/213) in the SHAP-repo. For anyone interested, [this](https://github.com/AndreCNF/shap) SHAP-fork was specifically made for RNNs/LSTMs, but it is quite old. I am unsure of how I need to format the background argument / inputs for the DeepExplainer. If I get to a working solution I will post an update.

**Reply by mbechtold** [reply]
Hi @LennardLux, did you eventually work on the SHAP method with PyTorch models? We have the same goal right now and were wondering whether somebody developed a tool that could be shared in the community. Thanks!

**Reply by LennardLux** [reply]
Hi @mbechtold, unfortunately I haven't been able to pursue my project any further since December last year due to time constraints.
